pkg/config/parameters.go

```diff
@@ -32,6 +32,7 @@ type ServerType string
 
 const ContourServerType ServerType = "contour"
 const EnvoyServerType ServerType = "envoy"
+const ContourGatewayClass = "projectcontour.io/contour"
```

Comment:
I wonder if we can use the class to know if we should dynamically provision Envoy or not? This isn't part of this review per-say just thinking about how to best manage that.

Reply:
In an effort to minimize the size of this PR, I'm adding Envoy support in a future PR. IMO the gc.controller string should simply differentiate implementations from one another. From my understanding, that's the intent of the field. By default, Contour should manage the Envoy infra when Gateway CRDs/CRs exist and the gc controller is "projectcontour.io/contour". If a user does not want Contour to manage the Envoy infra, we can add a config knob to the resource referenced by the gc paramRef. I prefer that we avoid adding this knob until someone makes a strong case for it.

Reply:
> I prefer that we avoid adding this knob until someone makes a strong case for it.

I've got a strong case for it, which was part of the reason moving this into Contour. I think we absolutely need to have a solution for folks who want a managed Envoy and for those who don't. 

The controller name is just a string, just thinking of using that to tell us which controller should manage stuff, meaning should Contour provision or not. It can go in the spec somewhere as well, just trying to reduce the number of fields that need configuration.

Reply:
The other thing to think about is how do we determine if a GatewayClass should be operator managed or not? Right now it will act on all instances, but do all users want that? The class could then be `projectcontour.io/operator` telling us that it's an operator managed instance vs a self-managed.

Reply:
@stevesloka IMO the proper way to provide the "--enableEnvoyMgt" config knob is through `gc.spec.parametersRef`. At first, the referent can be the contour.yaml configmap that includes the knob. I think it's beneficial in the future to transition the configmap to a structured object similar to what's being done with the operator.

Reply:
I think that we should avoid magic strings in the `gatewayClass.spec.controller` field as much as possible. That string should only be human-meaningful, not parsed to determine program options like which controller should use a particular GatewayClass. Configuring things with a field, either in a configmap or an explicit structure, is much more sustainable.

Reply:
> Configuring things with a field, either in a configmap or an explicit structure, is much more sustainable

@youngnick my thinking is to not make the gatewayclass controller identifier, i.e. `gc.spec.controller`, configurable. Instead, the ID is based on scheme "<API_GROUP>/<NAMESPACE>/contour". I don't understand the benefit of exposing this to a user.

Reply:
I was just following the docs:

> Controller is a domain/path string that indicates the controller that is managing Gateways of this class.

So we're defining which controller is the one in charge. We're introducing three different controller who are responsible. Is it the operator? Is it contour? Is it the cluster admin? Etc, etc.

Reply:
@stevesloka IMO the upstream docs need add'l details. I'm trying to define a `gc.spec.controller` naming schema for the gatewayclass contrller implemented in Contour. We have to consider that multiple Contour's may be running in the cluster, each with a gc controller. Therefore, `gc.spec.controller` should take the form of "CONTOUR_NAMESPACE/contour". I was originally thinking "CONTOUR_API_GROUP/CONTOUR_NAMESPACE/contour" but I reduced the verbosity based on https://github.com/projectcontour/contour/pull/3659#discussion_r633747716.

> We're introducing three different controller who are responsible. Is it the operator? Is it contour? Is it the cluster admin?

The number is based on the number of Contours in the cluster. Each instance of Contour will have its own `gc.spec.controller`. The operator will continue to support its existing controller string until we can deprecate Gateway API support in the operator.

Reply:
So is the goal @danehans to try and make the gatewayclass unique? So regardless of what the string looks like, we only have a problem if there are two gateways which reference the same gatewayclass (regardless of the controller)? Is that correct?

Reply:
@stevesloka that is correct. When Contour supports managing multiple Envoy data-planes, we can drop the existing Gateway ns/name from the Contour config. For every gateway that references the expected gatewayclass name, the matching Contour will deploy the envoy infra (daemonset, service, etc.).